protocols/dcutr/examples/client.rs

```diff
@@ -186,69 +186,75 @@ fn main() -> Result<(), Box<dyn Error>> {
+                SwarmEvent::Behaviour(Event::Identify(IdentifyEvent::Sent { .. })) => {
+                    info!("Told relay its public address.");
+                    told_relay_observed_addr = true;
                 }
-                SwarmEvent::Behaviour(Event::Identify(IdentifyEvent::Sent { .. })) => {}
                 SwarmEvent::Behaviour(Event::Identify(IdentifyEvent::Received {
                     info: IdentifyInfo { observed_addr, .. },
                     ..
                 })) => {
-                    info!("Relay observes us under the address: {:?}", observed_addr);
+                    info!("Relay told us our public address: {:?}", observed_addr);
                     learned_observed_addr = true;
                 }
                 event => panic!("{:?}", event),
             }
 
-            // Check whether we are done.
-
-            if !learned_observed_addr {
-                continue;
-            }
-
-            if opts.mode == Mode::Listen && !relay_accepted_reservation {
-                continue;
+            if learned_observed_addr && told_relay_observed_addr {
+                break;
             }
-
-            break;
         }
     });
 
-    if opts.mode == Mode::Dial {
-        swarm
-            .dial(
-                opts.relay_address
-                    .clone()
-                    .with(Protocol::P2pCircuit)
-                    .with(Protocol::P2p(opts.remote_peer_id.unwrap().into())),
-            )
-            .unwrap();
+    match opts.mode {
+        Mode::Dial => {
+            swarm
+                .dial(
+                    opts.relay_address
+                        .clone()
+                        .with(Protocol::P2pCircuit)
+                        .with(Protocol::P2p(opts.remote_peer_id.unwrap().into())),
+                )
+                .unwrap();
+        }
+        Mode::Listen => {
+            swarm
+                .listen_on(opts.relay_address.clone().with(Protocol::P2pCircuit))
+                .unwrap();
+
+            // Wait for relay to accept reservation request.
+            block_on(async {
+                loop {
+                    match swarm.next().await.unwrap() {
+                        SwarmEvent::NewListenAddr { .. } => {}
+                        SwarmEvent::Dialing { .. } => {}
+                        SwarmEvent::ConnectionEstablished { .. } => {}
+                        SwarmEvent::Behaviour(Event::Ping(_)) => {}
+                        SwarmEvent::Behaviour(Event::Relay(
+                            client::Event::ReservationReqAccepted { .. },
+                        )) => {
+                            info!("Relay accepted our reservation request.");
+                            break;
+                        }
+                        SwarmEvent::Behaviour(Event::Identify(_)) => {}
+                        event => panic!("{:?}", event),
+                    }
+                }
+            });
```

Comment:
Why do we need to do a separate loop here to wait for the `ReservationReqAccepted`? We don't initiate any following action, but instead directly jump into the next loop below. Couldn't we just remove this block?

Reply:
Good idea!